diff --git a/src/server.c b/src/server.c
index e6859cb..ecfe401 100644
--- a/src/server.c
+++ b/src/server.c
@@ -1407,11 +1407,11 @@ new_server(int fd, listen_ctx_t *listener)
     crypto->ctx_init(crypto->cipher, server->e_ctx, 1);
     crypto->ctx_init(crypto->cipher, server->d_ctx, 0);
 
-    int timeout = max(MIN_TCP_IDLE_TIMEOUT, server->listen_ctx->timeout);
+    int request_timeout = listener->timeout + rand() % 30;
     ev_io_init(&server->recv_ctx->io, server_recv_cb, fd, EV_READ);
     ev_io_init(&server->send_ctx->io, server_send_cb, fd, EV_WRITE);
     ev_timer_init(&server->recv_ctx->watcher, server_timeout_cb,
-                  timeout, timeout);
+                  request_timeout, listener->timeout);
 
     cork_dllist_add(&connections, &server->entries);
 
