diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 9ca9d85..09a417d 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -65,6 +65,7 @@ set(SS_SERVER_SOURCE
         cache.c
         resolv.c
         server.c
+		crypto-xor.c
         ${SS_CRYPTO_SOURCE}
         ${SS_PLUGIN_SOURCE}
         ${SS_ACL_SOURCE}
diff --git a/src/Makefile.am b/src/Makefile.am
index c255b60..ffacc48 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -31,6 +31,7 @@ crypto_src = crypto.c \
              aead.c \
              stream.c \
              ppbloom.c \
+			crypto-xor.c \
              base64.c
 
 plugin_src = plugin.c
@@ -109,7 +110,7 @@ libshadowsocks_libev_la_LDFLAGS = -version-info $(VERSION_INFO)
 libshadowsocks_libev_la_LIBADD = $(ss_local_LDADD)
 include_HEADERS = shadowsocks.h
 
-noinst_HEADERS = acl.h crypto.h stream.h aead.h json.h netutils.h redir.h server.h uthash.h \
+noinst_HEADERS = acl.h crypto.h crypto-xor.h stream.h aead.h json.h netutils.h redir.h server.h uthash.h \
                  cache.h local.h plugin.h resolv.h tunnel.h utils.h base64.h ppbloom.h \
                  common.h jconf.h manager.h rule.h socks5.h udprelay.h winsock.h
 EXTRA_DIST = ss-nat
diff --git a/src/common.h b/src/common.h
index 66e985c..a38ab95 100644
--- a/src/common.h
+++ b/src/common.h
@@ -31,6 +31,9 @@
 #endif
 
 #include "crypto.h"
+#ifdef MODULE_TUNNEL
+#include "netutils.h"
+#endif
 
 int init_udprelay(const char *server_host, const char *server_port,
 #ifdef MODULE_LOCAL
diff --git a/src/crypto.c b/src/crypto.c
index bd544b9..093a6df 100644
--- a/src/crypto.c
+++ b/src/crypto.c
@@ -42,6 +42,7 @@
 #include "aead.h"
 #include "utils.h"
 #include "ppbloom.h"
+#include "crypto-xor.h"
 
 int
 balloc(buffer_t *ptr, size_t capacity)
@@ -192,6 +193,23 @@ crypto_init(const char *password, const char *key, const char *method)
             memcpy(crypto, &temp, sizeof(crypto_t));
             return crypto;
         }
+
+        if (strncmp(method, "xor", 3) == 0) {
+            cipher_t *cipher = xor_init(password, NULL, NULL);
+            crypto_t *crypto = (crypto_t *)ss_malloc(sizeof(crypto_t));
+            crypto_t tmp     = {
+                .cipher      = cipher,
+                .encrypt_all = &xor_encrypt_all,
+                .decrypt_all = &xor_decrypt_all,
+                .encrypt     = &xor_encrypt,
+                .decrypt     = &xor_decrypt,
+                .ctx_init    = &xor_ctx_init,
+                .ctx_release = &xor_ctx_release,
+            };
+            memcpy(crypto, &tmp, sizeof(crypto_t));
+            return crypto;
+        }
+
         for (i = 0; i < STREAM_CIPHER_NUM; i++)
             if (strcmp(method, supported_stream_ciphers[i]) == 0) {
                 m = i;
diff --git a/src/utils.c b/src/utils.c
index 9241547..f310dcb 100644
--- a/src/utils.c
+++ b/src/utils.c
@@ -311,7 +311,7 @@ usage()
     printf(
         "       -k <password>              Password of your remote server.\n");
     printf(
-        "       -m <encrypt_method>        Encrypt method: plain, rc4-md5, \n");
+        "       -m <encrypt_method>        Encrypt method: plain, xor, rc4-md5, \n");
     printf(
         "                                  aes-128-gcm, aes-192-gcm, aes-256-gcm,\n");
     printf(
